### PR TITLE
Track Events: Fix template part track events not firing

### LINKED ISSUE #62924: Editor Tracks Events: "wpcom_block_editor_template_part_choose_existing" and "wpcom_block_editor_template_part_replace" no longer firing in site editor
### Quick summary

Two Tracks events we monitor in the site editor are...
- wpcom_block_editor_template_part_choose_existing --> fires when adding a Template Part block and selecting "Choose"
- wpcom_block_editor_template_part_replace --> fires when you select a Template Part block and click replace (and pick a new part)

On March 3, 2022, these seemed to stop firing.

Given that [Gutenberg 12.7.0](https://github.com/WordPress/gutenberg/releases/tag/v12.7.0) was deployed to simple sites on that day, it is most likely that Gutenberg upgrade broke our event handlers.

### Steps to reproduce

1. Go to site editor
2. Add a template part block
3. Click "Choose" and pick a template part
4. Click "Replace" and pick a new one
5. Check Tracks Live viewer -- you won't see the two events mentioned above show up!

### What you expected to happen

The two Tracks events fire.

### What actually happened

The Tracks events don't fire.

### Context

_No response_

### Simple, Atomic or both?

Simple, Atomic

### Theme-specific issue?

_No response_

### Browser, operating system and other notes

_No response_

### Reproducibility

Consistent

### Severity

Most (> 50%)

### Available workarounds?

There is no user impact

### Workaround details

No user impact, but our reporting data is off now.

### PR DESCRIPTION
**Pull request by edanzer**
### Context

Fixes https://github.com/Automattic/wp-calypso/issues/62924.

If you go the Site Editor, select a template, and add or replace a template part (via the template part block), there are separate Tracks events that should fire for each of those actions: 

- wpcom_block_editor_template_part_choose_existing --> fires when adding a Template Part block and selecting "Choose"
- wpcom_block_editor_template_part_replace --> fires when you select a Template Part block, click replace, and pick a new part

Those events have stopped firing. The reason is that [Gutenberg 12.7.0](https://github.com/WordPress/gutenberg/releases/tag/v12.7.0) updated the UI and thus selectors that our tracking code relied on. 

In addition, that release also consolidated the UI for both actions above. Before, the UI and selectors were different. Now they use the exact same components, so you can no longer differentiate the two actions above based on selector. 

### Proposed Changes

1) I've consolidated the tracking related code for the two events from two files into a single file. Because these events now rely on the same selector with conditional logic, it was DRY-er to manage them together.

2) I've updated selectors for both events to reflect the new UI. Note that both events now rely on the same selector. 

3) I've added conditional logic that determines which of the two events will fire. A capture event handlers runs prior to any changes and records details of template part block in `replaceData`. Later, when we run the tracking code, if that value of replaceData is undefined, it means we're inserting a new template part; if it is defined, it means we're replacing a template part.

### Testing Instructions

Here's a video walk through of what everything should look like once fixed:

https://user-images.githubusercontent.com/21228350/180883682-64cd19d1-2d39-4de6-b1dd-339a05b97d20.mov

Note: the affected code here is part of the wpcom-block-editor. So when testing, you'll need to cd into the wp-calypso/apps/wpcom-block-editor directory and use the `yarn dev --sync` command. 

**Duplicate the issue.** You may want to duplicate the issue first. To see the issue you can do the following. 
1. Checkout the trunk branch of calypso (and the wpcom-editor-plugin). Be sure to run `yarn install` from the the calypso root and `yarn dev --sync` from the wpcom-block-editor directory. 
2. Open the Site Editor and select any template to edit.
3. Click the + sign to add a block and select the Template Part block.
4. At this point, open your dev tools, go to the Network tab, and clear all entries so you can monitor new network activity. Also in the search/filter box, type `t.gif`. This will filter network activity so you can see Tracks events.
5. In the Template Part block, click the "Choose" button and clic  an available template part to insert.
6. In the network tab of you dev tools, note that you do not see a Tracks event. **This is the bug.**
7. In the same template part block, click the "Replace" button and click on a new template part. 
8. In the network tab of you dev tools, note that you do not see a Tracks event. **This is the bug.**

**Confirm the fix.** Now checkout the fix/template-part-tracking branch, and repeat steps 1-8 above.
- In step 6, you should now see a Tracks event in the network tab of your dev tools. Click on the event, look at the Payload, and look for the _en property. You should see wpcom_block_editor_template_part_choose_existing. 
- In step 8, you now see the wpcom_block_editor_template_part_replace event. As an extra check, confirm that that the replaced_template_part_id property shows the old template part you replaced, and that the template_part_id property shows the new template part you've selected.

### CODE REVIEW

**Comment by jeyip** [comment]
> In step 8, you should see two Tracks events in your network tab.

Should I still be seeing two events? I only see one, which is what I'd expect based off of our conversations on how to proceed with this PR

## Testing
### Requirements
- [x] `wpcom_block_editor_template_part_choose_existing` tracks event is sent when choosing a new template part
- [x]  `wpcom_block_editor_template_part_replace` tracks event is _not sent_ when choosing a new template part
- [x] `wpcom_block_editor_template_part_choose_existing` tracks event is _not sent_ when replacing an existing template part
- [x]  `wpcom_block_editor_template_part_replace` tracks event is sent when replacing an existing template part

File: apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-template-part-choose.js [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+ */
+export const wpcomTemplatePartChooseBubble = () => ( {
+	id: 'wpcom-block-editor-template-part-choose-existing',
+	selector:
+		'.block-library-template-part__selection-content .block-editor-block-patterns-list__item',
+	type: 'click',
+	handler: () => {
+		// If replaced_variation_slug is not defined, we're inserting a new template part.
+		if ( ! replaceData.replaced_variation_slug ) {
+			console.log( 'inserting new' ); // eslint-disable-line
```
Comment (jeyip) [inline apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-template-part-choose.js]: We can remove this console log

File: apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-template-part-choose.js [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+ */
+export const wpcomTemplatePartReplaceBubble = () => ( {
+	id: 'wpcom-block-editor-template-part-replace-bubble',
+	selector:
+		'.block-library-template-part__selection-content .block-editor-block-patterns-list__item',
+	type: 'click',
+	handler: () => {
+		// If replaced_variation_slug is defined, we're replacing a template part.
+		if ( replaceData.replaced_variation_slug ) {
+			console.log( 'replacing' ); // eslint-disable-line
```
Comment (jeyip) [inline apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-template-part-choose.js]: We can remove this console log

File: apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-template-part-choose.js (hunk 1 above)
Comment (edanzer) [reply apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-template-part-choose.js]: Just removed this. Thanks!

File: apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-template-part-choose.js (hunk 2 above)
Comment (edanzer) [reply apps/wpcom-block-editor/src/wpcom/features/tracking/wpcom-template-part-choose.js]: Removed!

**Comment by edanzer** [comment]
> > In step 8, you should see two Tracks events in your network tab.
> 
> Should I still be seeing two events? I only see one, which is what I'd expect based off of our conversations on how to proceed with this PR

@jeyip - Sorry, missed updating this line in testing instructions after our convo. I've updated it now. To confirm you should now only see ONE event there, which is the replace event.

**Review by jeyip — APPROVED** [review]
We discussed two approaches to solving the problem of differentiating between replacing an existing template part and choosing a new one.

1. Use the template part block attributes slug to determine whether or not we're inserting a new template part
2. Use the presence of the template part "Replace" button in the DOM tree to decide whether or not we're replacing a template part

In this PR, we go with approach #1, and although it's noticeably more complex, I prefer it because it's less on brittle and doesn't rely on yet another classname selector ( which, as we've seen, can change without warning ). 

+1 from me, but let's leave this for a day or two before merging so other folks have a chance to chime in

**Review by Addison-Stavlo — APPROVED** [review]
> In this PR, we go with approach https://github.com/Automattic/wp-calypso/issues/1, and although it's noticeably more complex, I prefer it because it's less on brittle and doesn't rely on yet another classname selector ( which, as we've seen, can change without warning ).

also +1 here.

This tests well on my end and the logic makes sense!

**Pull request merged**